### PR TITLE
feat(Core/Pets): Management refactoring

### LINKED ISSUE #3679: server is frozen [$150]
logs:
```
[Warning] Transaction aborted. 4 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 2450 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 7175 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 4527 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 4267 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 2450 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 6976 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 6297 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 499 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 499 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): DELETE FROM character_pet WHERE id = 133
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 6297 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 7175 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 6297 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 5219 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 4 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 6297 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 5219 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 4 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 7175 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 3 queries not executed.
SQL(p): UPDATE character_pet SET slot = 100 WHERE owner = 208 AND slot = 0
 [ERROR]: [1213] Deadlock found when trying to get lock; try restarting transaction
[Warning] Transaction aborted. 4 queries not executed.
```

 version:AzerothCore rev. c32283fab782+ 2020-11-06 18:05:03 +0000 (master branch) (Unix, Release) 
OS:ubuntu 18.04.4

<bountysource-plugin>

---
There is a **[$20 open bounty](https://www.bountysource.com/issues/94338996-server-is-frozen?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #9297: Resurrected Hunter Pets Stuck in Bones Form After Corpse Explosion
### What client do you play on?

enUS

### Faction

- [ ] Alliance
- [X] Horde

### Content Phase:

- [ ] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [X] 50-59

### Current Behaviour

Original report: https://github.com/chromiecraft/chromiecraft/issues/2372

Pet was turned into pile of bones in Scholomance. Pet was able to attack and use all abilities but appeared as a pile of bones. 

WoWScrnShot_112221_161447
WoWScrnShot_112221_161447

Tester: Confirmed this can happen. It's caused by the Scholomance Necromancer's [Corpse Explosion](https://tbc.wowhead.com/spell=17616/corpse-explosion) when used on a dead pet. If you rez the pet, it remains stuck in the bones form.

Pet in bones form, otherwise behaving normally:

https://user-images.githubusercontent.com/81782124/143024250-76d784a1-3b9c-4a2b-975f-1b14f5d2d2fa.mp4

### Expected Blizzlike Behaviour

The resurrected pet should return to its usual form.

### Source

Consistency says the resurrected pet shouldn't be stuck in pile-of-bones form.

### Steps to reproduce the problem

Go into scholomance with a hunter and pet (I had a wolf). Let the pet get killed and wait until a Necromancer casts Corpse Explosion on it. Then get out of combat and rez it before it fades away, and it will get stuck in bones form. The video linked below said you could take the bone-form pet outside the dungeon, but I wasn't able to duplicate that.

### Extra Notes

@Shineslippers found a video documenting how to reproduce this, but it's in Chinese, and is from Warmane rather than retail. It's here: https://www.youtube.com/watch?v=t-r8sdvUg8I

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/00dc369cb627d728c520cd597b31e9edf2c75eed

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Preload basic pet data on character login with async query
- Remove all select queries related to pet stable/unstable
- Add check pet before ressurect
- Set native display id after ressurect

## Issues Addressed:
- Closes #3679
- Closes #9297
- Part #9630

## SOURCE:
- Based TC commit - https://github.com/TrinityCore/TrinityCore/commit/5c1fc5e3876549f5ed2b9051fffb6f3d94a67d7a
- Continue #7306

## Tests Performed:
- Add/Delete
- Rename
- Swap slot
- Save/loading
- Combat
- Spells/Action bars
- #9589
- #9297

## How to Test the Changes:
- Just test all behavior pets

## Known Issues and TODO List:
- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Winfidonarleyan** [comment]
Need to test this PR very well

File: src/server/game/Entities/Pet/Pet.cpp [hunk 1]
```diff
@@ -2108,234 +2343,6 @@ void Pet::SynchronizeLevelWithOwner()
...
-    }
-
-    SetAsynchLoadType(PET_LOAD_DEFAULT);
-
-    // Warlock pet exception, check if owner is casting new summon spell
-    if (Spell* spell = owner->GetCurrentSpell(CURRENT_GENERIC_SPELL))
-        if (spell->GetSpellInfo()->HasEffect(SPELL_EFFECT_SUMMON_PET))
-            CastSpell(this, 32752, true, nullptr, nullptr, GetGUID());
-
-    if (owner->NeedSendSpectatorData() && GetCreatureTemplate()->family)
```
Comment (UltraNix) [inline src/server/game/Entities/Pet/Pet.cpp:2203]: That code is not longer used.

File: src/server/game/Entities/Pet/Pet.cpp [hunk 2]
```diff
@@ -2108,234 +2343,6 @@ void Pet::SynchronizeLevelWithOwner()
...
-
-    if (GetAsynchLoadType() == PET_LOAD_HANDLE_UNSTABLE_CALLBACK)
-    {
-        if (Player* player = owner->ToPlayer())
-            player->GetSession()->SendStableResult(0x09 /*STABLE_SUCCESS_UNSTABLE*/);
-    }
-    else// if (GetAsynchLoadType() == PET_LOAD_BG_RESURRECT || GetAsynchLoadType() == PET_LOAD_SUMMON_PET || GetAsynchLoadType() == PET_LOAD_SUMMON_DEAD_PET)
-    {
-        // Remove Demonic Sacrifice auras (known pet)
-        Unit::AuraEffectList const& auraClassScripts = owner->GetAuraEffectsByType(SPELL_AURA_OVERRIDE_CLASS_SCRIPTS);
```
Comment (UltraNix) [inline src/server/game/Entities/Pet/Pet.cpp:2125]: What about that code? Not longer needed?

File: src/server/game/Entities/Pet/Pet.cpp [hunk 3]
```diff
@@ -2108,234 +2343,6 @@ void Pet::SynchronizeLevelWithOwner()
...
-                        SetHealth(GetMaxHealth());
-
-                    if (GetMap()->IsBattleground())
-                        CastSpell(this, SPELL_PREPARATION, true);
-                }
-    }
-
-    // Fix aurastate auras, depending on health!
-    // Set aurastate manualy, prevents aura switching
-    if (HealthBelowPct(20))
```
Comment (UltraNix) [inline src/server/game/Entities/Pet/Pet.cpp:2173]: These changes to aura states not longer needed?

**Comment by temperrr** [comment]
Tested (with logout during the testing of each individual item on the list)

- [x] Untame/tame
- [x] pet create
- [x] Rename\incorrect rename
- [x] Swap slot
- [x] Save/loading
- [x] feed pet
- [x] Combat
- [x] Spells/Action bars
- [x] Hunter's talent Improved Revive Pet doesn't apply additional HP if pet corpse is despawned 
- [ ] Resurrected Hunter Pets Stuck in Bones Form After Corpse Explosion  -> doesnt fix this
image

File: src/server/game/Entities/Pet/Pet.cpp [hunk 4]
```diff
@@ -2388,3 +2395,40 @@ void Pet::RemoveSpellCooldown(uint32 spell_id, bool update /* = false */)
...
+    petInfo->PetNumber = m_charmInfo->GetPetNumber();
+    petInfo->CreatureId = GetEntry();
+    petInfo->DisplayId = GetNativeDisplayId();
+    petInfo->Level = getLevel();
+    petInfo->Experience = GetUInt32Value(UNIT_FIELD_PETEXPERIENCE);
+    petInfo->ReactState = GetReactState();
+    petInfo->Name = GetName();
+    petInfo->WasRenamed = !HasByteFlag(UNIT_FIELD_BYTES_2, 2, UNIT_CAN_BE_RENAMED);
+    petInfo->Health = GetHealth();
+    petInfo->Mana = GetPower(POWER_MANA);
```
Comment (UltraNix) [inline src/server/game/Entities/Pet/Pet.cpp:2410]: Not all pets use mana.

**Review by UltraNix — CHANGES_REQUESTED** [review]

**Comment by Winfidonarleyan** [comment]
@temperrr thanks for test. @UltraNix thanks for review

File: src/server/game/Entities/Pet/Pet.cpp (hunk 3 above)
Comment (Winfidonarleyan) [reply src/server/game/Entities/Pet/Pet.cpp:2173]: https://github.com/azerothcore/azerothcore-wotlk/blob/08d71d38d576ac75ef23b6df3bec56668b631726/src/server/game/Entities/Unit/Unit.cpp#L479

**Comment by Winfidonarleyan** [comment]
@temperrr i fix #9297, can retest?

**Comment by temperrr** [comment]
Tested:

- [x] #9297

**Comment by Winfidonarleyan** [comment]
> 

thanks

File: src/server/game/Entities/Pet/Pet.cpp (hunk 4 above)
Comment (Winfidonarleyan) [reply src/server/game/Entities/Pet/Pet.cpp:2410]: This default power

File: src/server/game/Entities/Pet/Pet.cpp (hunk 1 above)
Comment (Winfidonarleyan) [reply src/server/game/Entities/Pet/Pet.cpp:2203]: fixed

**Review by acidmanifesto — APPROVED** [review]

**Review by IntelligentQuantum — APPROVED** [review]

**Comment by Viste** [comment]
Маладец! одобряю!

**Review by Kitzunu — APPROVED** [review]

**Review by UltraNix — APPROVED** [review]

**Comment by mpfans** [comment]
> 

ok, I will test. If there are any bugs, I will submit the issue separately

**Pull request merged**